File: scripts/cli/coreutils-reference.py
"""DEV GNU reference capture/verify. Linux + bubblewrap only; no host fallback.

Each case runs twice in a fresh sandbox with no network or developer files.
Only system binaries/libraries, this worker and its request are read-only.
Normal --verify never changes evidence. --capture and --replace are explicit.
"""
import argparse
import hashlib
import json
import os
from pathlib import Path, PurePosixPath
import platform
import shutil
import subprocess
import sys
import tempfile
from datetime import datetime, timezone

ROOT = Path('/') if sys.argv[1:2] == ['--worker'] else Path(__file__).resolve().parents[2]
LOCK_PATH = 'content/cli-compatibility/coreutils-environment.json'
HARNESS_PATH = 'scripts/cli/coreutils-reference.py'


def canonical(value):
    return json.dumps(value, sort_keys=True, ensure_ascii=False, separators=(',', ':'))


def sha(data):
    return hashlib.sha256(data).hexdigest()


def source_hash(path):
    return sha((ROOT / path).read_text(encoding='utf-8').replace('\r\n', '\n').encode())


def request(case):
    # Same projection as referenceRequest in coreutils-reference.mjs.
    return {'id': case['id'], 'command': case['command'],
            'invocation': case.get('invocation', '/usr/bin/' + case['command']),
            'argv': case['argv'], 'stdinHex': case.get('stdinHex', (case.get('stdin') or '').encode().hex()),
            'env': case['env'], 'cwd': case['cwd'], 'fixture': {k: case.get('fixture', {}).get(k, [] if k in ['directories', 'setup'] else {}) for k in ['files', 'bytes', 'directories', 'modes', 'setup']},
            'process': case.get('process'), 'transport': case.get('transport', 'direct')}


def limits():
    import resource
    resource.setrlimit(resource.RLIMIT_FSIZE, (16 * 1024 * 1024, 16 * 1024 * 1024))
    resource.setrlimit(resource.RLIMIT_CPU, (4, 4))
    resource.setrlimit(resource.RLIMIT_AS, (512 * 1024 * 1024, 512 * 1024 * 1024))


def fixture_path(root, name):
    path = PurePosixPath(name)
    try:
        parts = path.relative_to('/home/kali').parts
    except ValueError as error:
        raise ValueError('Reference fixture paths must be within /home/kali') from error
    if '..' in parts or not parts:
        raise ValueError('Invalid fixture path')
    return root.joinpath(*parts)


def snapshot(root):
    result, groups = {}, {}
    for path in sorted(root.rglob('*')):
        st = path.lstat()
        row = {'inodeGroup': groups.setdefault(st.st_ino, len(groups)), 'mode': st.st_mode & 0o7777, 'nlink': st.st_nlink}
        if path.is_symlink():
            row.update(type='symlink', target=os.readlink(path))
        elif path.is_file():
            row.update(type='file', contentHex=path.read_bytes().hex(), size=st.st_size)
        else:
            row.update(type='directory')
        result[str(path.relative_to(root))] = row
    return result


def environment(req):
    process = req.get('process') or {}
    if process.get('environment') is not None:
        entries = process['environment']
    else:
        user = process.get('actor', 'kali')
        env = {'HOME': '/root' if user == 'root' else '/home/kali', 'HOSTNAME': 'lifeos',
               'LANG': 'C.UTF-8', 'LOGNAME': user,
               'PATH': '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
               'PWD': req['cwd'], 'SHELL': '/bin/bash', 'USER': user, **req['env']}
        entries = [[k, v] for k, v in sorted(env.items())]
    env = dict(entries)
    if env.get('LC_ALL') != 'C':
        raise ValueError('Every GNU case must explicitly use LC_ALL=C')
    return env


def worker(path):
    req = json.loads(Path(path).read_text())
    env = environment(req)
    argv = [req['invocation'], *req['argv']]
    if req['transport'] == 'redirect':
        fd = os.open('/home/kali/reference-output', os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o666)
        os.dup2(fd, 1)
        os.close(fd)
    elif req['transport'] == 'pipe':
        reader, writer = os.pipe()
        pid = os.fork()
        if pid:
            os.close(writer)
            while data := os.read(reader, 4096):
                view = memoryview(data)
                while view:
                    view = view[os.write(1, view):]
            os.close(reader)
            _, status = os.waitpid(pid, 0)
            os._exit(os.waitstatus_to_exitcode(status))
        os.close(reader)
        os.dup2(writer, 1)
        os.close(writer)
    binary = shutil.which(req['command'], path='/usr/bin:/bin')
    if not binary:
        raise RuntimeError('Missing locked GNU executable')
    os.execve(binary, argv, env)


def capture(case, bwrap):
    if case.get('script') or case['fixture'].get('setup') or case.get('inputEvents'):
        raise ValueError(f"{case['id']}: scripts/setup/events are project-only; use structured transport/identity")
    req = request(case)
    environment(req)
    process = req.get('process') or {}
    uid = process.get('uid', 1000)
    username = process.get('username', process.get('actor', 'kali'))
    with tempfile.TemporaryDirectory(prefix='cyber-war-coreutils-') as tmp:
        sandbox = Path(tmp)
        sandbox.chmod(0o755)
        home, etc = sandbox / 'home', sandbox / 'etc'
        home.mkdir(mode=0o777)
        home.chmod(0o777)
        etc.mkdir()
        (etc / 'passwd').write_text('' if username is None else f'{username}:x:{uid}:{uid}::/home/kali:/bin/sh\n')
        (etc / 'group').write_text(f'kali:x:{uid}:\n')
        for path in req['fixture']['directories']:
            fixture_path(home, path).mkdir(parents=True, exist_ok=True)
        for field in ['files', 'bytes']:
            for path, data in req['fixture'][field].items():
                target = fixture_path(home, path)
                target.parent.mkdir(parents=True, exist_ok=True)
                target.write_bytes(bytes.fromhex(data) if field == 'bytes' else data.encode())
        for path, mode in req['fixture']['modes'].items():
            fixture_path(home, path).chmod(mode)
        request_file = sandbox / 'request.json'
        request_file.write_text(canonical(req), encoding='utf-8')
        before = snapshot(home)
        cmd = [bwrap, '--unshare-all', '--die-with-parent', '--new-session', '--uid', str(uid), '--gid', str(uid)]
        for path in ['/usr', '/bin', '/lib', '/lib64']:
            if Path(path).exists():
                cmd += ['--ro-bind', path, path]
        cmd += ['--proc', '/proc', '--dev', '/dev', '--tmpfs', '/tmp',
                '--bind', str(home), '/home/kali', '--ro-bind', str(etc), '/etc',
                '--ro-bind', str(Path(__file__).resolve()), '/reference/worker.py',
                '--ro-bind', str(request_file), '/reference/request.json',
                '--chdir', req['cwd'], '--clearenv', '--setenv', 'LC_ALL', 'C',
                '--', '/usr/bin/python3', '/reference/worker.py', '--worker', '/reference/request.json']
        with tempfile.TemporaryFile() as stdout, tempfile.TemporaryFile() as stderr:
            completed = subprocess.run(cmd, input=bytes.fromhex(req['stdinHex']), stdout=stdout, stderr=stderr,
                                       timeout=5, check=False, preexec_fn=limits, env={'PATH': '/usr/bin:/bin', 'LC_ALL': 'C'})
            stdout.seek(0)
            stderr.seek(0)
            out, err = stdout.read(4 * 1024 * 1024 + 1), stderr.read(65537)
            if len(out) > 4 * 1024 * 1024 or len(err) > 65536:
                raise RuntimeError('Reference output limit exceeded')
            if completed.returncode < 0 or err.startswith(b'bwrap:') or b'Traceback (most recent call last)' in err:
                raise RuntimeError(f'Sandbox/worker failed for {case["id"]}: {err[:1000]!r}')
        return {'id': req['id'], 'request': req, 'requestDigest': sha(canonical(req).encode()),
                'stdoutHex': out.hex(), 'stderrHex': err.hex(), 'exitCode': completed.returncode,
                'before': before, 'after': snapshot(home)}


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    action = parser.add_mutually_exclusive_group(required=True)
    action.add_argument('--capture', action='store_true')
    action.add_argument('--verify', action='store_true')
    parser.add_argument('--replace', action='store_true')
    scope = parser.add_mutually_exclusive_group(required=True)
    scope.add_argument('--command')
    scope.add_argument('--wave', choices=['foundation'])
    parser.add_argument('--output-dir')
    parser.add_argument('--probe', help='DEV inputs; writes observations outside approved goldens')
    args = parser.parse_args()
    if args.replace and not args.capture:
        parser.error('--replace requires --capture')
    if sys.platform != 'linux' or not shutil.which('bwrap'):
        parser.error('DEV capture requires Linux and bubblewrap; Windows gameplay has no such dependency')
    manifest = json.loads((ROOT / 'content/cli-compatibility/manifest.json').read_text())
    version = manifest['software']['coreutils']['referenceVersion']
    config = json.loads((ROOT / 'content/cli-compatibility/coreutils.json').read_text())
    lock = json.loads((ROOT / LOCK_PATH).read_text())
    if lock['coreutilsVersion'] != version:
        parser.error('Reference environment baseline mismatch')
    names = [args.command] if args.command else [k for k, v in config['commands'].items() if v['area'] == args.wave]
    if any(n not in config['commands'] for n in names):
        parser.error('Unknown executable')
    cases = json.loads(Path(args.probe).read_text()) if args.probe else sum([json.loads(p.read_text()) for p in sorted((ROOT / 'tests/cli/compat/pilot').glob('*.json'))], [])
    cases = [c for c in cases if c['softwareId'] == 'coreutils' and c['command'] in names]
    if not cases:
        parser.error('No selected cases')
    binaries = {}
    for name in sorted({c['command'] for c in cases}):
        located = shutil.which(name, path='/usr/bin:/bin')
        if not located:
            parser.error(f'Missing GNU executable: {name}')
        binary = Path(located)
        result = subprocess.run([str(binary), '--version'], capture_output=True, timeout=5, check=True, env={'LC_ALL': 'C'})
        first = result.stdout.decode().splitlines()[0]
        if first != f'{name} (GNU coreutils) {version}':
            parser.error(f'Reference version mismatch: {first}; expected {version}')
        binaries[name] = {'versionLine': first, 'sha256': sha(binary.read_bytes())}
        if name in lock.get('binaryHashes', {}) and binaries[name]['sha256'] != lock['binaryHashes'][name]:
            parser.error(f'Locked GNU binary hash mismatch: {name}')
    output = Path(args.output_dir) if args.output_dir else ROOT / f'tests/cli/gnu/coreutils/{version}'
    if args.probe and not args.output_dir:
        parser.error('Probes require a separate --output-dir')
    env = {'id': lock['id'], 'lockHash': source_hash(LOCK_PATH), 'os': 'Linux', 'osRelease': Path('/etc/os-release').read_text(),
           'architecture': platform.machine(), 'isolation': 'bubblewrap/unshare-all', 'network': 'disabled',
           'locale': 'C', 'binaryHashes': binaries, 'packages': subprocess.check_output(['/sbin/apk', 'info', '-v'], stderr=subprocess.DEVNULL, text=True).splitlines()}
    for name in names:
        selected = [c for c in cases if c['command'] == name]
        target = output / f'{name}.json'
        if args.capture and target.exists() and not args.replace:
            parser.error(f'{target} exists; explicit --replace required')
        rows = []
        for case in selected:
            first, second = capture(case, shutil.which('bwrap')), capture(case, shutil.which('bwrap'))
            if first != second:
                raise RuntimeError(f'Non-reproducible GNU reference: {case["id"]}')
            first['reproducible'] = True
            rows.append(first)
        payload = {'schemaVersion': 2, 'provenance': 'GNU_PROBE' if args.probe else 'GNU_REFERENCE', 'version': version,
                   'command': name, 'locale': 'C', 'capturedAt': datetime.now(timezone.utc).isoformat(),
                   'harnessHash': source_hash(HARNESS_PATH), 'environment': {**env, 'binaryHashes': {name: binaries[name]}}, 'cases': rows}
        if args.verify:
            original = json.loads(target.read_text())
            for key in ['schemaVersion', 'provenance', 'version', 'command', 'locale', 'harnessHash', 'environment', 'cases']:
                if original[key] != payload[key]:
                    raise RuntimeError(f'Reference reproduction differs: {name}/{key}; golden unchanged')
        else:
            output.mkdir(parents=True, exist_ok=True)
            target.write_text(json.dumps(payload, indent=2, ensure_ascii=False) + '\n', encoding='utf-8')
        print(f'{name}: {len(rows)} GNU {version} cases, two identical runs; {"verified" if args.verify else "captured"}')


if __name__ == '__main__':
    if sys.argv[1:2] == ['--worker']:
        worker(sys.argv[2])
    else:
        main()

File: scripts/cli/coreutils-reference.mjs
// DEV evidence validation. This module only reads captures; it never invokes GNU.
import { createHash } from 'node:crypto';
import { existsSync } from 'node:fs';
import { resolve } from 'node:path';
import { json, read, root } from './io.mjs';
export const canonical = (v) =>
  JSON.stringify(
    v && typeof v === 'object'
      ? Array.isArray(v)
        ? v.map((x) => JSON.parse(canonical(x)))
        : Object.fromEntries(
            Object.keys(v)
              .filter((k) => v[k] !== undefined)
              .sort()
              .map((k) => [k, JSON.parse(canonical(v[k]))]),
          )
      : v,
  );
const hash = (s) => createHash('sha256').update(s).digest('hex');
export const sourceHash = (p) => hash(read(p).replaceAll('\r\n', '\n'));
export function referenceRequest(c) {
  const request = {
    id: c.id,
    command: c.command,
    invocation: c.invocation ?? `/usr/bin/${c.command}`,
    argv: c.argv,
    stdinHex: c.stdinHex ?? Buffer.from(c.stdin ?? '').toString('hex'),
    env: c.env,
    cwd: c.cwd,
    fixture: Object.fromEntries(
      ['files', 'bytes', 'directories', 'modes', 'setup'].map((k) => [
        k,
        c.fixture?.[k] ?? (['directories', 'setup'].includes(k) ? [] : {}),
      ]),
    ),
    process: c.process ?? null,
    transport: c.transport ?? 'direct',
  };
  if (['cat', 'head', 'tail'].includes(c.command)) {
    request.io = c.io ?? null;
    request.interaction = c.interaction ?? null;
    for (const key of ['hardlinks', 'symlinks']) request.fixture[key] = c.fixture?.[key] ?? {};
  }
  return request;
}
export const requestDigest = (c) => hash(canonical(referenceRequest(c)));
export function referenceCapture(command) {
  const path = `tests/cli/gnu/coreutils/${command.referenceVersion}/${command.command}.json`;
  return existsSync(resolve(root, path)) ? json(path) : null;
}
export function validateReference(command, capture) {
  if (!capture) return 'No GNU reference capture';
  const binary = capture.environment?.binaryHashes?.[command.command];
  const lock = json('content/cli-compatibility/coreutils-environment.json');
  if (
    capture.schemaVersion !==
      (command.command === 'tail' ? 4 : ['cat', 'head'].includes(command.command) ? 3 : 2) ||
    capture.provenance !== 'GNU_REFERENCE' ||
    capture.version !== command.referenceVersion ||
    capture.command !== command.command ||
    capture.locale !== 'C' ||
    !Number.isFinite(Date.parse(capture.capturedAt))
  )
    return 'Reference provenance/baseline/schema/locale mismatch';
  if (
    capture.harnessHash !== sourceHash('scripts/cli/coreutils-reference.py') ||
    capture.environment?.lockHash !==
      sourceHash('content/cli-compatibility/coreutils-environment.json')
  )
    return 'Reference harness/environment fingerprint stale';
  if (
    ['cat', 'head', 'tail'].includes(command.command) &&
    capture.interactionHash !== sourceHash('scripts/cli/coreutils_interaction.py')
  )
    return 'Reference interaction harness fingerprint stale';
  if (
    command.command === 'tail' &&
    capture.followHash !== sourceHash('scripts/cli/coreutils_follow.py')
  )
    return 'Reference follow harness fingerprint stale';
  if (
    capture.environment?.os !== 'Linux' ||
    capture.environment?.id !== lock.id ||
    capture.environment?.architecture !== 'x86_64' ||
    capture.environment?.locale !== 'C' ||
    capture.environment?.network !== 'disabled' ||
    capture.environment?.isolation !== 'bubblewrap/unshare-all' ||
    !binary?.sha256?.match(/^[a-f0-9]{64}$/) ||
    binary.sha256 !== lock.binaryHashes?.[command.command] ||
    binary.versionLine !== `${command.command} (GNU coreutils) ${command.referenceVersion}`
  )
    return 'Reference binary/environment identity invalid';
  if (
    !Array.isArray(capture.cases) ||
    new Set(capture.cases.map((c) => c.id)).size !== capture.cases.length
  )
    return 'Duplicate/missing reference cases';
  return null;
}
export function referenceDifference(test, actual, row) {
  if (
    test.script ||
    test.fixture?.setup?.length ||
    test.inputEvents?.length ||
    test.roundtrip ||
    (test.tty?.isTTY && !test.interaction)
  )
    return 'GNU structured reference does not cover scripts/events/roundtrip/TTY execution';
  if (!row || row.reproducible !== true) return 'Missing reproducible GNU case';
  if (
    row.requestDigest !== requestDigest(test) ||
    canonical(row.request) !== canonical(referenceRequest(test))
  )
    return 'GNU request fingerprint mismatch';
  if (!actual) return 'No current project execution';
  const errors = [];
  for (const stream of ['stdout', 'stderr']) {
    const expected = row[stream + 'Hex'],
      observed = actual[stream + 'Hex'] ?? Buffer.from(actual[stream] ?? '').toString('hex');
    if (typeof expected !== 'string' || !/^(?:[a-f0-9]{2})*$/.test(expected)) {
      errors.push(`Invalid GNU ${stream} bytes`);
      continue;
    }
    if (expected !== observed) {
      const a = Buffer.from(expected, 'hex'),
        b = Buffer.from(observed, 'hex');
      let offset = 0;
      while (offset < a.length && offset < b.length && a[offset] === b[offset]) offset++;
      errors.push(
        `${stream} byte ${offset}: GNU=${a.subarray(offset, offset + 24).toString('hex')} project=${b.subarray(offset, offset + 24).toString('hex')}`,
      );
    }
  }
  if (row.exitCode !== actual.exitCode)
    errors.push(`status GNU=${row.exitCode} project=${actual.exitCode}`);
  if (['cat', 'head', 'tail'].includes(test.command)) {
    const io = test.io ?? {};
    const endpoints = {
      stdin:
        test.interaction?.schemaVersion === 2
          ? io.stdinPath
            ? 'file'
            : 'null'
          : test.interaction
            ? 'pty-canonical-echo-off'
            : io.stdinPath
              ? 'file'
              : 'pipe',
      stdout:
        io.stdoutPath || test.transport === 'redirect'
          ? 'file'
          : io.closedConsumer
            ? 'closed-pipe'
            : 'pipe',
      stderr: io.stderrPath ? 'file' : 'pipe',
    };
    if (canonical(row.endpoints) !== canonical(endpoints))
      errors.push('GNU descriptor provenance mismatch');
    const terminationStatus =
      row.termination?.kind === 'exit'
        ? row.termination.code
        : row.termination?.kind === 'signal'
          ? { SIGINT: 130, SIGPIPE: 141, SIGTERM: 143 }[row.termination.signal]
          : undefined;
    if (terminationStatus === undefined || terminationStatus !== row.exitCode)
      errors.push('Invalid GNU termination result');
    for (const field of ['termination', 'observations'])
      if (canonical(row[field]) !== canonical(actual[field]))
        errors.push(`${field} differs from GNU`);
    const nodes = actual.after?.vfs ?? {};
    for (const relative of Object.keys(row.before)) {
      if (!Object.hasOwn(row.after, relative) && Object.hasOwn(nodes, '/home/kali/' + relative))
        errors.push(`Path removed by GNU still exists: ${relative}`);
    }
    const pairs = [];
    for (const [relative, expected] of Object.entries(row.after)) {
      const path = '/home/kali/' + relative,
        observed = nodes[path];
      if (!observed) {
        errors.push(`Missing VFS path ${path}`);
        continue;
      }
      for (const key of ['mode', 'nlink'])
        if (observed[key] !== expected[key]) errors.push(`${path}/${key} differs from GNU`);
      if (observed.kind !== expected.type) errors.push(`${path}/type differs from GNU`);
      if (expected.type === 'file' && actual.files?.[path] !== expected.contentHex)
        errors.push(`${path}/bytes differ from GNU`);
      if (expected.type === 'symlink' && observed.content !== expected.target)
        errors.push(`${path}/target differs from GNU`);
      for (const [group, ino] of pairs)
        if ((group === expected.inodeGroup) !== (ino === observed.ino))
          errors.push(`${path}/inode relationship differs from GNU`);
      pairs.push([expected.inodeGroup, observed.ino]);
    }
    const stable = (nodes) =>
      Object.fromEntries(
        Object.entries(nodes)
          .filter(
            ([p]) =>
              (!Object.hasOwn(row.after, p.slice('/home/kali/'.length)) &&
                !Object.hasOwn(row.before, p.slice('/home/kali/'.length))) ||
              !p.startsWith('/home/kali/'),
          )
          .map(([p, n]) => [
            p,
            {
              kind: n.kind,
              content: n.content,
              blob: n.blob ?? null,
              mode: n.mode,
              ino: n.ino,
              nlink: n.nlink,
            },
          ]),
      );
    if (canonical(stable(actual.before.vfs)) !== canonical(stable(nodes)))
      errors.push('Unexpected project filesystem mutation');
    return errors.join('; ') || null;
  }
  if (test.transport === 'redirect') {
    const output = actual.after?.vfs?.['/home/kali/reference-output'];
    const bytes = output?.blob ? null : Buffer.from(output?.content ?? '').toString('hex');
    if (!output || bytes !== row.after?.['reference-output']?.contentHex)
      errors.push('Redirected VFS bytes differ from GNU');
    if (
      output?.mode !== row.after?.['reference-output']?.mode ||
      output?.nlink !== row.after?.['reference-output']?.nlink
    )
      errors.push('Redirected VFS metadata differs from GNU');
  } else if (canonical(row.before) !== canonical(row.after))
    errors.push('Unexpected GNU filesystem mutation');
  // Foundation has no filesystem writes beyond the declared stdout redirect.
  const before = actual.before?.vfs ?? {},
    after = actual.after?.vfs ?? {};
  const selected = (nodes) =>
    Object.fromEntries(
      Object.entries(nodes)
        .filter(([p]) => test.transport !== 'redirect' || p !== '/home/kali/reference-output')
        .map(([p, n]) => [
          p,
          {
            kind: n.kind,
            content: n.content,
            blob: n.blob ?? null,
            mode: n.mode,
            uid: n.uid,
            gid: n.gid,
            ino: n.ino,
            nlink: n.nlink,
          },
        ]),
    );
  if (canonical(selected(before)) !== canonical(selected(after)))
    errors.push('Unexpected project filesystem mutation');
  return errors.join('; ') || null;
}
